internal/proxy/task_delete.go

```diff
@@ -202,75 +202,78 @@ func repackDeleteMsgByHash(
 	collectionName string,
 	partitionID int64,
 	partitionName string,
-) (map[uint32]*msgstream.DeleteMsg, int64, error) {
+) (map[uint32][]*msgstream.DeleteMsg, int64, error) {
+	maxSize := Params.PulsarCfg.MaxMessageSize.GetAsInt()
 	hashValues := typeutil.HashPK2Channels(primaryKeys, vChannels)
 	// repack delete msg by dmChannel
-	result := make(map[uint32]*msgstream.DeleteMsg)
+	result := make(map[uint32][]*msgstream.DeleteMsg)
+	lastMessageSize := map[uint32]int{}
+
 	numRows := int64(0)
+	numMessage := 0
+
+	createMessage := func(key uint32, vchannel string) *msgstream.DeleteMsg {
+		numMessage++
+		lastMessageSize[key] = 0
+		return &msgstream.DeleteMsg{
+			BaseMsg: msgstream.BaseMsg{
+				Ctx: ctx,
+			},
+			DeleteRequest: &msgpb.DeleteRequest{
+				Base: commonpbutil.NewMsgBase(
+					commonpbutil.WithMsgType(commonpb.MsgType_Delete),
+					// msgid of delete msg must be set later
+					// or it will be seen as duplicated msg in mq
+					commonpbutil.WithTimeStamp(ts),
+					commonpbutil.WithSourceID(paramtable.GetNodeID()),
+				),
+				CollectionID:   collectionID,
+				PartitionID:    partitionID,
+				CollectionName: collectionName,
+				PartitionName:  partitionName,
+				PrimaryKeys:    &schemapb.IDs{},
+				ShardName:      vchannel,
+			},
+		}
+	}
+
 	for index, key := range hashValues {
 		vchannel := vChannels[key]
-		_, ok := result[key]
+		msgs, ok := result[key]
 		if !ok {
-			deleteMsg, err := newDeleteMsg(
-				ctx,
-				idAllocator,
-				ts,
-				collectionID,
-				collectionName,
-				partitionID,
-				partitionName,
-			)
-			if err != nil {
-				return nil, 0, err
-			}
-			deleteMsg.ShardName = vchannel
-			result[key] = deleteMsg
+			result[key] = make([]*msgstream.DeleteMsg, 1)
+			msgs = result[key]
+			result[key][0] = createMessage(key, vchannel)
+		}
+		curMsg := msgs[len(msgs)-1]
+		size, id := typeutil.GetId(primaryKeys, index)
+		if lastMessageSize[key]+16+size > maxSize {
+			curMsg = createMessage(key, vchannel)
+			result[key] = append(result[key], curMsg)
 		}
-		curMsg := result[key]
 		curMsg.HashValues = append(curMsg.HashValues, hashValues[index])
 		curMsg.Timestamps = append(curMsg.Timestamps, ts)
 
-		typeutil.AppendIDs(curMsg.PrimaryKeys, primaryKeys, index)
+		typeutil.AppendID(curMsg.PrimaryKeys, id)
+		lastMessageSize[key] += 16 + size
```

Comment:
Can we use `unsafe.Sizeof()` here?

Reply:
Too slow，will use sizeof function for every rows.